Add support for “intensity” modifiers on box and text shadows

thecrypticace opened the pull request:
This PR adds support for utilities like:
- `text-shadow-lg/25`

It uses relative color syntax to replace the alpha value of the shadow from your theme.

When combined with a colors:
- `text-shadow-lg/25 text-shadow-red-500`
- `text-shadow-lg/25 text-shadow-red-500/75`

The alpha values are **multiplied** resulting in a shadow with the color specified in `--color-red-500` and alpha values of 25% and 18.75% respectively.

thecrypticace committed:
Add support for “intensity” modifiers on text shadows

thecrypticace committed:
Add support for “intensity” modifiers on box shadows

thecrypticace committed:
Update tests

thecrypticace committed:
Update UI tests

thecrypticace committed:
Add UI tests

thecrypticace committed:
Add support for “intensity” modifiers on inset box shadows

thecrypticace committed:
Make sure `inset-shadow-none` is suggested

RobinMalfait commented on packages/tailwindcss/src/__snapshots__/index.test.ts.snap:
```diff
@@ -36,6 +36,12 @@ exports[`compiling CSS > \`@tailwind utilities\` is replaced by utilities using
 }
 
+@property --tw-shadow-intensity {
+  syntax: "*";
```
Should we use a concrete type here? 🤔 
```suggestion
  syntax: "<percentage>";
```

Not sure if this is required if you want to change the intensity on hover and have a (smooth) transition.

RobinMalfait reviewed (changes requested):
Left a few comments, I also wonder/wish we could only enable features like this _if_ you use them 🤔

thecrypticace committed:
Refactor

thecrypticace committed:
Inline intensity value

We don’t need to use the variable in these cases because the intensity is available directly to the utility when generated.

This also helps gurantee that the relative color syntax will only be present in development and when Lighting CSS is unable to inline it

thecrypticace committed:
Add more tests

thecrypticace committed:
Update data type

thecrypticace committed:
Update tests

thecrypticace commented on packages/tailwindcss/src/utilities.ts:
```diff
@@ -181,6 +181,23 @@ export function withAlpha(value: string, alpha: string): string {
+ * Apply opacity to a color using `color-mix`.
+ */
+export function replaceAlpha(value: string, alpha: string): string {
+  if (alpha === null) return value
```
I added the `| null` 👍

thecrypticace commented on packages/tailwindcss/src/utilities.ts:
```diff
@@ -4258,15 +4278,32 @@ export function createUtilities(theme: Theme) {
+              value,
+              (color) =>
+                `var(--tw-text-shadow-color, ${intensity ? replaceAlpha(color, 'var(--tw-text-shadow-intensity)') : color})`,
+            ),
```
Good idea. It also made me realize we don't need to replaceAlpha with the variable here. Just the literal `intensity` value (which when `null` does nothing).

Using the literal value also means that Lighting CSS should down-level this to non-relative-color syntax when possible.

RobinMalfait reviewed (approved):
Nice, I like it.

thecrypticace commented:
@adamwathan you cool with the name `--tw-shadow-intensity`? I could use `--tw-shadow-alpha` too if you'd prefer that.

thecrypticace committed:
Rename variable

thecrypticace committed:
Update tests

thecrypticace committed:
Update changelog
